File: src/Combinations.cpp
#include <Combinations.hpp>

using namespace std;

int Combinations::calculate_trips_combinations(Data &data)
{
    int nb_trains = data.get_nb_trains();
    int nb_routes = data.get_nb_routes();

    all_trips_combinations.resize(nb_trains);
    for (int i = 0; i < nb_trains; i++)
    {
        // verify whether trips were already calculated for the train
        bool skip = false;
        for (int j = i-1; j >= 0; j--)
        {
            if (data.get_train_max_trips(j) == data.get_train_max_trips(i))
            {
                all_trips_combinations[i] = all_trips_combinations[j];
                skip = true;
                break;
            }
        }
        if (!skip)
        {
            if (verify_overflow(nb_routes, data.get_train_max_trips(i)))
            {
                // error: too many combinations!
                return 0;
            }
            int nb_trips_comb_for_train = pow(nb_routes+1, data.get_train_max_trips(i)); // using nb_routes+1 to consider that no route is made
            
            // calculate number of trips combinations
            generate_trips_combinations(data, i);
        }
    }
    return 1;
}

void Combinations::generate_trips_combinations(Data &data, int train_idx)
{
    int nb_routes = data.get_nb_routes();
    int nb_trips_comb_for_train = pow(nb_routes+1, data.get_train_max_trips(train_idx));

    vector<int> current(data.get_train_max_trips(train_idx), 0);
    for (unsigned long long count = 0; count < nb_trips_comb_for_train; count++)
    {
        // verify feasibility of the current combination before adding to the vector
        if (check_trips_feasibility(data, current))
        {
            all_trips_combinations[train_idx].push_back(current);
        }

        // go to next combination
        for (int i = data.get_train_max_trips(train_idx)-1; i >= 0; i--)
        {
            if (++current[i] < nb_routes+1)
                break;
            current[i] = 0;
        }
    }
}

bool Combinations::verify_overflow(unsigned long long base, unsigned long long exp)
{
    unsigned long long result = 1;
    for (unsigned long long i = 0; i < exp; i++)
    {
        if (result > ULLONG_MAX / base)
            return true;
        result *= base;
    }
    return false;
}

bool Combinations::is_valid_combination (Data &data, RoutesCombination &combination)
{
    if (!normalize_combination(data, combination))
    {
        return false; // return false if combination was already previously executed
    }

    // verify whether demands were met
    vector <int> demands_per_day (data.get_nb_vertices(), 0);
    for (int i = 0; i < data.get_nb_vertices(); i++)
    {
        for (int j = 0; j < data.get_nb_intervals(); j++)
            demands_per_day[i] += data.get_demands()[i][j];
    }

    vector<int> times_vertex_was_visited (data.get_nb_vertices(), 0);
    for (int i = 0; i < combination.size(); i++)
    {
        for (int j = 0; j < combination[i].size(); j++)                        
        {
            if (combination[i][j] != data.get_nb_routes() && combination[i][j] != -1)
            {
                int route = combination[i][j];
                for (auto vertex : data.get_route_vertices(route))
                { 
                    times_vertex_was_visited[vertex]++;
                }
            }
        }
    }
    for (int i = 0 ; i < data.get_nb_vertices(); i++)
    {
        if (times_vertex_was_visited[i] < data.get_demand_per_day()[i]) return false;
    }
    return true;
}


bool Combinations::check_trips_feasibility (Data &data, vector<int> &sequence_of_trips)
{
    int nb_routes = data.get_nb_routes();
    
    // verify whether first route starts at the initial depot
    if (sequence_of_trips[0] != nb_routes)
    {
        if (!data.is_valid_route(0, 0, sequence_of_trips[0])) return false;
    }

    bool flag = false; // flag to tell whether train completes any trips during the day
    for (int i = 0; i < sequence_of_trips.size()-1; i++)
    {
        if (sequence_of_trips[i] != nb_routes)
        {
            flag = true;
            if (sequence_of_trips[i+1] != nb_routes)
            {
                // verify whether subsequential routes are compatible
                if (data.are_incompatible_routes(sequence_of_trips[i], sequence_of_trips[i+1])) return false;
            }
        }
        else if (sequence_of_trips[i] == nb_routes)
        {
            // if trip is not made, verify whether the next ones also are not made
            if (i != sequence_of_trips.size()-1 && sequence_of_trips[i+1] != nb_routes) return false;
        }
    }

    if (sequence_of_trips.front() != nb_routes) flag = true;

    return flag; 
}

bool Combinations::normalize_combination (Data &data, RoutesCombination &combination)
{
    // normalize combinations to verify whether it was already tested previously
    // (only changes the train that will complete the trips)
    // return false if it was already tested and returns true if it's a new combination

    int nb_routes = data.get_nb_routes();
    // make sure all dimensions have the same size
    for (int i = 0; i < combination.size(); i++)
    {
        while (combination[i].size() < data.get_max_nb_trips())
            combination[i].push_back(nb_routes);
    }
    // normalize the vector
    vector<vector<int>> normalized_combination = combination;
    sort(normalized_combination.begin(), normalized_combination.end());

    // verify whether it was already added according to the result
    pair<set<vector<vector<int>>>::iterator, bool> result;
    #pragma omp critical
    {
        result = unique_combinations.insert(normalized_combination);
    }
    if (!result.second)
    {
        return false; // combination already exists
    }
    return true; // current combination is a new combination
}

File: python/combinations.py
import numpy as np
import math
from threading import Lock

ULLONG_MAX = (1 << 64) - 1

class Combinations:
    def __init__(self, data):
        self.data = data

        self.all_trips_combinations = None
        self.unique_combinations = set()
        self.lock = Lock()

    def verify_sequence_feasibility(self, seq, route_not_completed):
        flag = True
        last_idx = len(seq) - 1
        for i in range(last_idx):
            cur = seq[i]
            nxt = seq[i + 1]

            if cur != route_not_completed:
                # verify compatibility between consecutive routes
                if nxt != route_not_completed:
                    if self.data.are_incompatible_routes(cur, nxt):
                        return False
            else:
                # after a non-trip, no trips can exist
                if nxt != route_not_completed:
                    return False

        return flag
        
    def check_trips_feasibility(self, seq):

        # seq: represents a sequence of trips a single train can complete
        route_not_completed = self.data.nb_routes

        # verify whether first route starts at the initial depot
        if seq[0] != route_not_completed:
            if not self.data.is_valid_route(0, 0, seq[0]):
                return False
        else:
            # verify whether other routes are also not completed
            if not all(route == route_not_completed for route in seq[1:]):
                return False
            else:
                return True

        return self.verify_sequence_feasibility(seq, route_not_completed)

    def generate_trips_combinations(self, total, max_trips):

        current = np.zeros(max_trips, dtype=np.uint16)

        valid_list = []
        routes_plus_1 = self.data.nb_routes + 1
        for _ in range(total):
            if self.check_trips_feasibility(current):
                valid_list.append(current.copy())

            # go to next combination
            for pos in range(max_trips - 1, -1, -1):
                current[pos] += 1
                if current[pos] < routes_plus_1:
                    break
                current[pos] = 0

        return valid_list

    def calculate_trips_combinations(self):

        def verify_overflow(base, exp):
            if base <= 1:
                return False
            return exp * math.log(base) > math.log(ULLONG_MAX)

        self.all_trips_combinations = [None] * self.data.nb_trains
        computed = {}

        for i in range(self.data.nb_trains):
            max_trips = self.data.max_trips_per_train[i]

            if max_trips in computed:
                self.all_trips_combinations[i] = computed[max_trips]
                continue

            if verify_overflow(self.data.nb_routes + 1, max_trips):
                raise ValueError("Too many combiantions! Aborting...")

            total = (self.data.nb_routes + 1) ** max_trips
            self.all_trips_combinations[i] = self.generate_trips_combinations(total,max_trips)
            computed[max_trips] = self.all_trips_combinations[i]

    def normalize_combination(self, combination):
        nb_routes = self.data.nb_routes

        # ensure all dimensions have the same size
        max_trips = self.data.max_nb_trips
        for i in range(len(combination)):
            # convert numpy array to list if needed
            if isinstance(combination[i], np.ndarray):
                combination[i] = combination[i].tolist()
            while len(combination[i]) < max_trips:
                combination[i].append(nb_routes)

        # create a normalized copy
        normalized_combination = sorted((tuple(row) for row in combination))
        normalized_tuple = tuple(normalized_combination)

        # insert with thread safety
        with self.lock:
            if normalized_tuple in self.unique_combinations:
                return False  # already seen
            self.unique_combinations.add(normalized_tuple)
        return True  # a new combination

    def verify_daily_demands(self, combination):

        times_vertex_was_visited = [0] * self.data.get_nb_vertices()
        for i in range(len(combination)):
            for j in range(len(combination[i])):
                if combination[i][j] == -1: # if route is undefined cannot calculate if demands were met
                    return True
                if combination[i][j] != self.data.nb_routes:
                    route = combination[i][j]
                    for vertex in self.data.route_vertices[route]:
                        times_vertex_was_visited[vertex] += 1

        # if any demand was not met, invalid combination
        for i in range(self.data.get_nb_vertices()):
            if times_vertex_was_visited[i] < self.data.demand_per_day[i]:
                return False

        return True

    def is_valid_combination(self, combination):

        if not self.normalize_combination(combination):
            return False

        if not self.verify_daily_demands(combination):
            return False

        return True



